Second approach for #4295

claudioandre-br opened the pull request:
See `git diff leakage leak2`

I'm not sure how `reset()` exactly works. It seems, this is a better approach (I mean, any format must call `release_clobj` inside `create_clobj`).

Commit:
OpenCL: fix a leak in SHA256 and SHA512 (raw and crypt)

Make the source code of the different formats more similar.
Must create bitmap in init() (raw: SHA256/SHA512)

Now that we are releasing the OpenCL buffers, we must force the creation
of the bitmap in init().

ghost commented:
This PR fixes all leaks in my formats.

magnumripper commented on src/opencl_rawsha256_fmt_plug.c:
```diff
@@ -165,6 +165,9 @@ static void create_clobj(size_t gws, struct fmt_main *self)
 
+	// Avoid memory leaks.
+	release_clobj();
+
```
The above probably needs to be added to all other (not yours) formats as well, right?

claudioandre-br replied:
Yes, using bash-fu (if possible). I use a fancy IDE.

magnumripper replied:
What IDE, Eclipse?

claudioandre-br replied:
Visual Studio Code

solardiz commented on src/opencl_rawsha256_fmt_plug.c:
```diff
@@ -299,7 +300,9 @@ static void release_clobj()
+		ocl_initialized = 0;
+		HANDLE_CLERROR(clFinish(queue[gpu_id]), "Error releasing memory");
+
 	}
```
Extra empty line before the closing curly brace. Drop it.

solardiz commented on src/opencl_rawsha256_fmt_plug.c:
```diff
@@ -165,6 +165,9 @@ static void create_clobj(size_t gws, struct fmt_main *self)
 	uint32_t hash_id_size;
 	size_t mask_cand = 1, mask_gws = 1;
 
+	// Avoid memory leaks.
```
I think these comments don't help, and should be dropped. It is implied that if we're freeing resources, we're doing so not to leak them.

solardiz reviewed (approved):
I made some nitpick comments - feel free to make changes per them or not.

magnumripper commented:
I added a commit fixing all other formats (bar a few strange Sayantan ones I don't grok). Unfortunetely I added that comment to all of them, lol.

solardiz commented:
> Unfortunetely I added that comment to all of them, lol.

I don't understand this remark. Are you saying some formats have duplicate logic after that commit - e.g., both Claudio's changes and yours?

magnumripper commented:
> > Unfortunetely I added that comment to all of them, lol.
> 
> I don't understand this remark. Are you saying some formats have duplicate logic after that commit - e.g., both Claudio's changes and yours?

No I added the redundant `// Avoid memory leaks` comment - but it's gone now due to some perl-fu and a force-push

solardiz reviewed (approved):
The 3 commits' combined changes look good to me now.

magnumripper commented:
I'm happy with this now. No leaks seen, even in RAR 😵

ghost commented:
I'll check it later.
```
Testing: o5logon-opencl, Oracle O5LOGON protocol [SHA1 AES OpenCL]... *** Error in `../run/john': malloc(): memory corruption: 0x00007f83929b5710 ***
```
```
Testing: o5logon-opencl, Oracle O5LOGON protocol [SHA1 AES OpenCL]... 0: OpenCL CL_INVALID_MEM_OBJECT (-38) error in opencl_o5logon_fmt_plug.c:143 - Error Unmapping result
```

solardiz commented:
@claudioandre-br Are these o5logon-opencl failures new with this PR's changes or did they occur before? What system/device are you seeing them on?

solardiz commented:
Oh, I see these are on Travis. So it's something we must look into before merging this.

Commit:
OpenCL formats (many); Fix memory leaks

Fix a bug in o5logon introduced recently. MEM_FREE nulls the pointer,
OpenCL functions don't.
Fix RAW formats invalid pointer
Handle release_clobj call even if not initialized
Ensure buffers are always destroyed before re-created

See #4295

ghost commented:
We can squash commits:
2a9e64fb8b29fd3ed65d71b6fb774e96a83464de
56e9bcdea7956591e5d9930f6ff12f2ead8960d0
ba4fdba843b32dfd7ab70b66b1adc469f16829f6

If you guys think this is mandatory, I will create a new PR (because I need to keep this branch to compare).

solardiz commented:
Now Travis failed on:

```
Testing: raw-MD4-opencl [MD4 OpenCL]... 0: OpenCL CL_INVALID_VALUE (-30) error in opencl_hash_check_128_plug.c:374 - failed in clEnqueueWriteBuffer buffer_bitmaps.
```

After fixing whatever other issues there might be, I think we should squash all commits here (currently all 6 of them). We can do that via GitHub UI if we don't mind GitHub appearing as the committer in git history, or @claudioandre-br can create a new PR for that as he suggested.

ghost commented:
IMO, these commits are not generic and should be maintained.
c6ae3795ec1e5c09bd57126e570e718eccdc6c07 
1832a5b5a9f74bd38a4d6314be2fb3bd85c6e9fa

magnumripper commented:
I see no particular reason to squash at all

magnumripper commented:
...but I'm not merging this now if that's a concern, please do at will

magnumripper commented on src/opencl_rawsha1_fmt_plug.c:
```diff
@@ -192,6 +192,7 @@ static void release_clobj_kpc(void)
 		HANDLE_CLERROR(clReleaseMemObject(pinned_saved_idx), "Error Releasing pinned_saved_idx.");
 		HANDLE_CLERROR(clReleaseMemObject(pinned_int_key_loc), "Error Releasing pinned_int_key_loc.");
 		buffer_idx = 0;
+		pinned_saved_keys = 0;
```
I'd have set them as NULL, not that it makes any difference

solardiz replied:
Sure, let's use NULL for pointers.

solardiz replied:
I see we still use 0 and not NULL here. These are `cl_mem`, so it's non-obvious whether any such usage is correct at all - maybe that was meant to be an opaque type for which we don't reliably know whether it's an integer, pointer, or something else? On a given system I see it's `typedef struct _cl_mem *            cl_mem;`, so maybe NULL is more correct, but ideally we'd know more about what's guaranteed (or not) about this type and use these variables only as guaranteed to work by OpenCL spec.

solardiz commented:
> IMO, these commits are not generic and should be maintained.
> [c6ae379](https://github.com/openwall/john/commit/c6ae3795ec1e5c09bd57126e570e718eccdc6c07)
> [1832a5b](https://github.com/openwall/john/commit/1832a5b5a9f74bd38a4d6314be2fb3bd85c6e9fa)

As I understand, the first one of these not only fixes but also introduces now-known bugs, and the second one fixes those. I'd squash.

solardiz commented:
Claudio, if you don't want the 6 commits squashed into one, please create a new PR with clean commits where we don't have a commit introduce known bugs nor even comments that we end up deleting later in the PR. Each commit should be reasonable on its own (at least not introducing new known bugs). While at it, please also take care of magnum's comment about NULLs. Thanks!

ghost commented:
I tested (up to 100 iterations):
- rawMd5-OpenCL (still leaking; probably, the author of the format missed some buffer, maybe program[], queue[], ...
- rawSHA256-OpenCL (Ok)
- rawSHA512-OpenCL (Ok)
- xSHA512-OpenCL (Ok)
- SHA256Crypt-OpenCL (Ok)
- SHA512Crypt-OpenCL (Ok)
****
BTW: I'm testing on NVIDIA 1080. Is it safe, right? Or it'll burn?

solardiz commented:
> BTW: I'm testing on NVIDIA 1080. Is it safe, right? Or it'll burn?

It's fine, but I suggest moving to the Titan X.

ghost commented:
> It's fine, but I suggest moving to the Titan X.

I like to run these tests on 1080 (how the leak evolves). But I will at least let it rest.

ghost commented:
Final thoughts: all formats need to be audited (later). #4295 should be kept open.

solardiz commented:
> I will at least let it rest.

This is unnecessary. It's just that you get a severely throttled clock rate after only a few minutes on this poor card. If you only care about memory and not speeds, and you need to compare against the reference memory usage figures you saw earlier, you can simply continue to use the card like you've been using it.

magnumripper commented:
I've tried to find the leak in raw-md5-opencl. Beats me so far.

Notes
After first test round, 168 MiB is lost. That's after a GWS of 16384. Then after each more round w/ same size, another 157 MiB is lost.

magnumripper commented:
OK so I used this code with `--stress-test` option:
```diff
diff --git a/src/bench.c b/src/bench.c
index fa3191976..20a7c9262 100644
--- a/src/bench.c
+++ b/src/bench.c
@@ -732,6 +732,9 @@ int benchmark_all(void)
 		puts("NOTE: This is a debug build, speed will be lower than normal");
 #endif
 
+	fprintf(stderr, "GPU memory at start: ");
+	if (system("nvidia-smi --query-gpu=memory.used --format=csv,noheader"))
+		error();
 AGAIN:
 	options.loader.field_sep_char = 31;
 #endif
@@ -1110,6 +1114,8 @@ next:
 		}
 #ifndef BENCH_BUILD
 	if (options.flags & FLG_LOOPTEST) {
+		if (system("nvidia-smi --query-gpu=memory.used --format=csv,noheader"))
+			error();
 		if (event_abort) {
 			uint32_t p = 100 * loop_fail / loop_total;
 			uint32_t pp = 10000 * loop_fail / loop_total - p * 100;
```

And these debug macros:
```diff
diff --git a/src/opencl_common.h b/src/opencl_common.h
index 727f91731..76b610caf 100644
--- a/src/opencl_common.h
+++ b/src/opencl_common.h
@@ -287,6 +287,16 @@ void opencl_process_event(void);
 		} \
 	} while (0)
 
+/* Debug functions */
+#define jtrCreateBuffer(buf, context, flags, size, wat, ret)	  \
+	do { \
+		if ((buf)) \
+			error_msg("** %s:%d Memory leak\n", __FILE__, __LINE__); \
+		(buf) = clCreateBuffer((context), (flags), (size), (wat), (ret)); \
+	} while (0)
+
+#define jtrReleaseMemObject(buf) { cl_mem _b = (buf); (buf) = NULL; clReleaseMemObject(_b); }
+
 /* Macro for get a multiple of a given value */
 #define GET_NEXT_MULTIPLE(dividend, divisor)	  \
 	(divisor) ? ((dividend + (ocl_v_width * divisor - 1)) / (ocl_v_width * divisor)) * divisor : (dividend) / ocl_v_width;
```

I replaced ALL instances of creating or releasing buffers in opencl_rawmd5_fmt_plug.c and opencl_hash_check_128_plug.c - there simply is no leak there. I verified that if I commented out some clReleaseMemObject line, my macros catched it.

Still:
```
GPU memory at start: 205 MiB
Device 1: GeForce RTX 2080 Ti
#1 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=32 GWS=16384 DONE
Raw:    7166M c/s real, 7237M c/s virtual, Dev#1 util: 100%

373 MiB
#2 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=64 GWS=32768 DONE
Raw:    14333M c/s real, 14333M c/s virtual, Dev#1 util: 99%

530 MiB
#3 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=64 GWS=32768 DONE
Raw:    14333M c/s real, 14263M c/s virtual, Dev#1 util: 99%

687 MiB
#4 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=64 GWS=32768 DONE
Raw:    14333M c/s real, 14263M c/s virtual, Dev#1 util: 99%

844 MiB
#5 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=64 GWS=32768 DONE
Raw:    14254M c/s real, 14183M c/s virtual, Dev#1 util: 99%

1001 MiB
```

Obviously I've also had a good look at alloc/calloc/free (EDIT I now added assertions for all of them, no problem seen). Also, the format appears to destroy mappings, kernel and program correctly but maybe I should write more debug assertions like the above one in opencl_common.h.

EDIT 2: Same leak seen without auto-tune.

EDIT 3: Same leak seen with `--test --no-mask`

magnumripper commented:
Just self-test without benchmark does not leak!
```
$ ../run/john --stress-test=0 -form:raw-md5-opencl 
GPU memory at start: 205 MiB
Device 1: GeForce RTX 2080 Ti
#1 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
#2 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
#3 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
```
Same for higher LWS/GWS (I'm not certain it's honored though, it used to be)
```
$ ../run/john --stress-test=0 -form:raw-md5-opencl -lws=64 -gws=4194304
GPU memory at start: 205 MiB
Device 1: GeForce RTX 2080 Ti
#1 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
#2 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
#3 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
```
EDIT: The higher GWS was **not** honored.

ghost commented:
> OK so I used this code with --stress-test option:

Me likes some debug code seen here. Very useful.

magnumripper committed:
OpenCL: Canned function-like macros for chasing leaks

See #4354

magnumripper commented:
> Me likes some debug code seen here. Very useful.

I added them in a separate commit - I think we should keep them.

magnumripper commented:
Running those debug functions and `--format=opencl`, it looks like we're leaking a little even though I don't get any detected misses in buffer releases. We should probably add similar debug macros for `cl(Create|Release)Kernel()` and `cl(Create|Release)Program()`.

EDIT: Did that. Now it looks more like GPU memory is "slow" to recover, or fragmented. Any given single format will not "leak" while running all of them will look like they do, a little. Note also that many formats use `mem_alloc_tiny` somewhere which in itself is a memory leak when running this way - but usually not IRL.

magnumripper commented:
DEScrypt-opencl doesn't even have a `done()` 👀

magnumripper commented:
I get segfaults from mscash2-opencl, disabling it from more tests. Sayantan again, it's a pity he's not with us now

magnumripper commented:
Got this now
```
#2 Testing: raw-SHA256-opencl [SHA256 OpenCL]... 0: OpenCL CL_INVALID_KERNEL (-48) error in opencl_rawsha256_fmt_plug.c:150 - Error setting argument 6
```
It's reproducible alone as `../run/john --stress-test=0 -form:raw-sha256-opencl`. My debugging stuff was in place. EDIT: It doesn't have to be.

magnumripper commented:
> DEScrypt-opencl doesn't even have a `done()` 👀

I'm opening a separate issue for that one. It's not a trivial one to fix, I always hated trying to follow the strange code paths through different files.

magnumripper commented:
After 35dd7fc, I can run all formats (except mscash2-opencl that just segfaults and descrypt-opencl which has #4362, I refuse to look at them) several times with no detected leak. I think this PR is ready to merge.

magnumripper commented:
> After [35dd7fc](https://github.com/openwall/john/commit/35dd7fc22d31294e0118e06202d46039bff51327), I can run all formats (except mscash2-opencl that just segfaults and descrypt-opencl which has #4362, I refuse to look at them) several times with no detected leak. I think this PR is ready to merge.

No wait, I only ran through `--stress-test=0` several rounds. I obviously need to do a `--stress-test` a couple rounds also.

ghost commented:
> except mscash2-opencl [...] and descrypt-opencl 

To test, we need to remove hard to maintain formats. :+1:

magnumripper commented:
I see now many formats have this helper macro
```
#define CLCREATEBUFFER(_flags, _size, _string) \
   clCreateBuffer(context[gpu_id], _flags, _size, NULL, &cl_error); \
   HANDLE_CLERROR(cl_error, _string);
```
My previous perl-fu for patching in the debug functions would miss them

solardiz commented:
> except mscash2-opencl that just segfaults

@magnumripper Please create an issue for the mscash2-opencl segfault, including reproduction instructions. Thanks!

solardiz commented:
> To test, we need to remove hard to maintain formats. +1

But we need to keep those formats available for use. I think mscash2-opencl is fairly easy to rewrite in a manner consistent with other formats, and that should be done (we'll lose this format's unique ability/hack to use multiple devices without forking, but that's OK). descrypt-opencl is much harder.

magnumripper commented:
I just add them to disabled formats section in my `john-local.conf`, no problem.

solardiz commented:
> I think this PR is ready to merge.

I don't like having the 13 separate commits. I think the separation between some of them is more historical (during the 3-day history of this PR) rather than logical. Ideally, you'd squash some of the commits manually, having fewer properly logically separated commits to merge, and with titles reflecting their differences from other commits here. Would you do that, @magnumripper? Thanks!

solardiz commented on src/bench.c:
```diff
@@ -10,9 +10,6 @@
  */
 
-#define _BSD_SOURCE /* for setenv() */
-#define _DEFAULT_SOURCE 1 /* for setenv() */
```
Hmm. I've just confirmed there are no uses of `setenv` in `bench.c`. OK.

magnumripper replied:
Yeah I wondered where they went but didn't look further into it

solardiz commented on src/bench.c:
```diff
@@ -732,6 +731,12 @@ int benchmark_all(void)
 		puts("NOTE: This is a debug build, speed will be lower than normal");
 #endif
 
+	if (options.flags & FLG_LOOPTEST && system("which nvidia-smi >/dev/null") == 0) {
```
Maybe we can limit this to OpenCL formats? Ideally, also to uses of an NVIDIA device, but that's harder. Also, please add braces around `options.flags & FLG_LOOPTEST`.

solardiz commented on src/bench.c:
```diff
@@ -724,6 +721,8 @@ int benchmark_all(void)
 	int ompt_start = omp_get_max_threads();
 #endif
 	const char *opencl_was_skipped = "";
+	int nvidia_mem = 0;
```
Should probably be inside `#ifndef BENCH_BUILD`.

solardiz commented on src/bench.c:
```diff
@@ -1070,6 +1075,8 @@ int benchmark_all(void)
 
 next:
+		if (nvidia_mem && system("nvidia-smi --query-gpu=memory.used --format=csv,noheader"))
+			error();
```
Should probably be inside `#ifndef BENCH_BUILD`. And please test `make generic`.

solardiz commented on src/opencl_common.h:
```diff
@@ -287,6 +274,44 @@ void opencl_process_event(void);
+
+#define jtrReleaseMemObject(buf) ({ cl_mem _b = (buf); (buf) = NULL; clReleaseMemObject(_b); })
+#define jtrReleaseKernel(kernel) ({ cl_kernel _k = (kernel); (kernel) = NULL; clReleaseKernel(_k); })
+#define jtrReleaseProgram(program) ({ cl_program _p = (program); (program) = NULL; clReleaseProgram(_p); })
```
I think we should standardize on `john` in naming of related projects, identifiers in code, etc. Not `jtr`.

magnumripper replied:
Will try to remember that for #4363 (perhaps I'll end up using neither, just eg. OCLRELEASEMEM)

solardiz reviewed (approved):
I'm posting some review comments, even though I see you've already pushed more commits that I haven't included yet.

magnumripper commented:
> > I think this PR is ready to merge.

That was premature, still testing.

> I don't like having the 13 separate commits. I think the separation between some of them is more historical (during the 3-day history of this PR) rather than logical. Ideally, you'd squash some of the commits manually, having fewer properly logically separated commits to merge, and with titles reflecting their differences from other commits here. Would you do that, @magnumripper? Thanks!

Will do. I use git autosquash feature so my recent (strange) commit descriptions already tell git how to handle half of it when I later do a `git rebase -i`.

magnumripper commented:
I've had occasional segfaults from mysql-sha1-opencl and mscash-opencl as well now, but can't reproduce. It really looks like Sayantan did something in his code that either isn't quite compatible with our shared code, or just plain wrong in some other aspect or detail - virtually all such problems are with his few formats. It's just that I've spent hours and hours trying to find it over the years and I can't see any particular problem! I bet once we find it, it's a similar problem across all of them.

magnumripper commented:
I see things like notes-opencl segfaulting on an all-formats test (right after LM-opencl, who would have thought) but not on its own (we've seen that before). I'm actually considering adding option `--format-list=LIST` so I could test with `--format=lm-opencl,notes-opencl`. I'd use the existing list parsing of course, like we do with several other recent-ish options like `--single-seed=WORD[,WORD...]`

magnumripper commented:
Done testing now. I'll merge once bots are happy
